mypy/plugins/dataclasses.py

```diff
@@ -811,3 +844,43 @@ def _has_direct_dataclass_transform_metaclass(info: TypeInfo) -> bool:
         info.declared_metaclass is not None
         and info.declared_metaclass.type.dataclass_transform_spec is not None
     )
+
+
+def replace_function_sig_callback(ctx: FunctionSigContext) -> CallableType:
+    """
+    Returns a signature for the 'dataclasses.replace' function that's dependent on the type
+    of the first positional argument.
+    """
+    if len(ctx.args) != 2:
+        # Ideally the name and context should be callee's, but we don't have it in FunctionSigContext.
+        ctx.api.fail(f'"{ctx.default_signature.name}" has unexpected type annotation', ctx.context)
+        return ctx.default_signature
+
+    if len(ctx.args[0]) != 1:
+        return ctx.default_signature  # leave it to the type checker to complain
+
+    obj_arg = ctx.args[0][0]
+
+    # <hack>
+    from mypy.checker import TypeChecker
+
+    assert isinstance(ctx.api, TypeChecker)
+    obj_type = ctx.api.expr_checker.accept(obj_arg)
+    # </hack>
+
+    obj_type = get_proper_type(obj_type)
+    if not isinstance(obj_type, Instance):
+        return ctx.default_signature
+
+    replace_func = obj_type.type.get_method(_INTERNAL_REPLACE_SYM_NAME)
```

Comment:
I don't like this hack at all :(

> I wish I could've deserialized metadata, or even accessed a list of DataclassAttributes that's already been deserialized...

So, what's the problem? Looks like we have it in the metadata anyway:

```python
info.metadata["dataclass"] = {
            "attributes": [attr.serialize() for attr in attributes],
```

Reply:
Metadata should contain serialized JSON, not ad hoc type objects that I don't mean to serialize. That's why I'm saying I'd like if there'd be a way to stash runtime arbitrary data belonging to a plugin on a TypeInfo.

Reply:
You can deserialize attributes with https://github.com/python/mypy/blob/9944d5fc6ae29a862bfab980a42a9bfae89ee5c0/mypy/plugins/dataclasses.py#L149 in `replace_function_sig_callback` and do not store anything extra. 

Or am I missing something?

Reply:
I can look it up again, but I believe I couldn't do it with the API provided by that hook, but only with the semantic analysis API.

Also
- it makes sense to build the signature once and then cache it, and
- (more importantly IMO) the other signature, for `__init__`, is built ahead of time, so it'd be more consistent to build the `replace` signature at the same time and place

`replace` could've been an instance method and then we'd add it together with `__init__`. It being a static method is "coincidental".

BTW, I considered implementing it differently, by replacing `replace` with a synthetic overloaded func, then have the plugin append overloads, but I think it'd result in worse performance and worse error messages (imagine having an error listing of all `replace` "overloads"  for all dataclasses in your codebase).

Reply:
> it makes sense to build the signature once and then cache it,

Yes, sounds reasonable. But, other plugins do not do this. I don't think that this use-case should be the first one to do something like this.

> I considered implementing it differently, by replacing replace with a synthetic overloaded func

This can quickly get out of control, if you have a lot of arguments including optional ones.

Plus, I agree with your point that overloads can generate unreadable error messages.

> I can look it up again, but I believe I couldn't do it with the API provided by that hook, but only with the semantic analysis API.

I cannot see anything specific about it. So, the algorithm would be:

```python
def replace_sig_hook(ctx):
    attributes = deserialize_metadata()
    signature = compute_from_dataclass_attributes(attributes)
    return signature
```

Please, in case you have any problems - let me know, I would be happy to help!

In case this would be very slow (and users would complain) we can find ways to optimize it in the future 👍